korge/src/commonMain/kotlin/com/soywiz/korge/view/animation/ParallaxDataView.kt

```diff
@@ -0,0 +1,432 @@
+                            addUpdater {
+                                // Calculate the offset for the inner scrolling of the layer
+                                y += (deltaY * speedY + selfSpeedY) * it.milliseconds
+                            }
+                        }
+                        ParallaxMode.NO_PLANE -> {
+                            addUpdater {
+                                // Calculate the offset for the inner scrolling of the layer
+                                x += (deltaX * speedX + selfSpeedX) * it.milliseconds
+                                y += (deltaY * speedY + selfSpeedY) * it.milliseconds
+                            }
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    init {
+        // Only the base container for all view objects needs to be scaled
+        this.scale = scale
+
+        // First create background layers in the back
+        constructLayer(data.backgroundLayers, data.config.backgroundLayers, data.config.mode, smoothing)
+
+        // Then construct the two parallax planes with their attached layers
+        if (data.config.mode != ParallaxMode.NO_PLANE) {
+            constructParallaxPlane(
+                data.parallaxPlane,
+                data.attachedLayers,
+                data.config.parallaxPlane,
+                data.config.mode == ParallaxMode.HORIZONTAL_PLANE,
+                smoothing
+            )
+            // Do horizontal or vertical movement depending on parallax scrolling direction
+            if (data.config.mode == ParallaxMode.HORIZONTAL_PLANE) {
+                // Move parallax plane inside borders
+                addUpdater {
+                    // Sanity check of diagonal movement - it has to be between 0.0 and 1.0
+                    diagonal = diagonal.clamp(0.0, 1.0)
+                    y = -(diagonal * (parallaxLayerSize - data.config.size.height))
+                }
+            } else {
+                addUpdater {
+                    diagonal = diagonal.clamp(0.0, 1.0)
+                    x = -(diagonal * (parallaxLayerSize - data.config.size.width))
+                }
+            }
+        }
+
+        // Last construct the foreground layers on top
+        constructLayer(data.foregroundLayers, data.config.foregroundLayers, data.config.mode, smoothing)
+    }
+}
+
+suspend fun VfsFile.readParallaxDataContainerFromAseFile(
+    config: ParallaxConfig,
+    atlas: MutableAtlas<Unit>? = null
+): ParallaxDataContainer {
+    val props = ImageDecodingProps(this.baseName, extra = ExtraTypeCreate())
+    return ParallaxDataContainer(
+        config = config,
+        backgroundLayers = if (config.backgroundLayers != null) {
+            props.setExtra("layers", config.backgroundLayers.joinToString(separator = ",") { it.name })
+            props.setExtra("disableSlicing", true)
+            val out = ASE.readImage(this.readAsSyncStream(), props)
```

Comment:
This line:

```
val out = ASE.readImage(this.readAsSyncStream(), props)
```

seems to be repeated a few times. It is reading the same ASE file?

Reply:
Yes, it reads the same ASE file. The reason is that I need at least once read the file without "slicing" enabled (see "disableSlicing" ImageDecodingProps) and once read the file with slicing enabled. Slicing is used for generating the parallax (3D) plane from the ASE file. The normal layer images (e.g. for the foreground layer) should not be sliced from the ASE file but just taken as the full image.
Then I was also easier handling the different layer types (background, foreground and attached layers) in the ParallaxDataView when they are in different ImageData objects. This is why I read it 4 times.

Reply:
I see. Thanks for the explanation!